mne/coreg.py

```diff
@@ -1242,3 +1245,572 @@ def _scale_xfm(subject_to, xfm_fname, mri_name, subject_from, scale,
+                self._nearest_transformed_high_res_mri_idx_hpi])
+            weights.append(np.full(len(mri_pts[-1]), self._hpi_weight))
+        head_pts = np.concatenate(head_pts)
+        mri_pts = np.concatenate(mri_pts)
+        weights = np.concatenate(weights)
+        if self._n_scale_param == 0:
+            mri_pts *= self._scale  # not done in fit_matched_points
+        return head_pts, mri_pts, weights
+
+    def fit_icp(self, n_iterations=20, lpa_weight=1., nasion_weight=10.,
+                rpa_weight=1.):
+        """Find MRI scaling, translation, and rotation to match HSP.
+
+        Parameters
+        ----------
+        n_iterations : int
+            Maximum number of iterations.
+        lpa_weight : float
+            Relative weight for LPA. The default value is 1.
+        nasion_weight : float
+            Relative weight for nasion. The default value is 10.
+        rpa_weight : float
+            Relative weight for RPA. The default value is 1.
+        """
+        self._lpa_weight = lpa_weight
+        self._nasion_weight = nasion_weight
+        self._rpa_weight = rpa_weight
+
+        # Initial guess (current state)
+        est = self._parameters
+        est = est[:[6, 7, None, 9][self._n_scale_param]]
+
+        # Do the fits, assigning and evaluating at each step
+        for _ in range(n_iterations):
+            head_pts, mri_pts, weights = self._setup_icp()
+            est = fit_matched_points(mri_pts, head_pts,
+                                     scale=self._n_scale_param,
+                                     x0=est, out='params', weights=weights)
+            if self._n_scale_param == 0:
+                self._update_params(est[:3], est[3:6])
+            elif self._n_scale_param == 1:
+                self._update_params(est, est[-1], est[-1])
+            else:
+                self._update_params(est[:3], est[3:6], est[6:9])
+            angle, move, scale = self._changes
+            if angle <= self._icp_angle and move <= self._icp_distance and \
+                    all(scale <= self._icp_scale):
+                break
+
+    def omit_hsp_points(self, distance):
+        """Exclude head shape points that are far away from the MRI head.
+
+        Parameters
+        ----------
+        distance : float
+            Exclude all points that are further away from the MRI head than
+            this distance. A value of distance <= 0 excludes nothing.
+        """
+        distance = float(distance)
+        if distance <= 0:
+            return
+
+        # find the new filter
+        mask = self._orig_hsp_point_distance <= distance
+        n_excluded = np.sum(~mask)
+        logger.info("Coregistration: Excluding %i head shape points with "
+                    "distance >= %.3f m.", n_excluded, distance)
+        # set the filter
+        with warnings.catch_warnings(record=True):  # comp to None in Traits
+            self._dig.extra_points_filter = mask
```

Comment:
This `omit_hsp_points` will need to call some kind of update. Maybe we should add a `force_update_omitted=False` to `self._update_params` and then this can call:
```
self._update_params(force_update_omitted=True)
```
And in the `_update_params` the filtered points will be updated as above (when things change) *or* when `force_update_omitted is True`.

This also made me realize that this class is missing a `self.set_grow_hair` method that exists in `mne coreg`. Basically it "pushes out" the MRI points. This method can just set the value and then do `self._update_params(self._rotation, self._translation, self._scale)` or so, because when the hair grows I think everything needs to be updated IIRC.

Reply:
Why do we have to recompute here? I thought the goal with `omit_hsp_points()` was to use the mask and hide the points in the GUI.

Reply:
Not just the GUI, it's used during fitting. Let's say you have 100 "extra" points but 3 were accidentally digitized 5 cm from the head (it happens sometimes if you mis-click the digitizing pen for example in Polhemus). You omit by distance and use the remaining points for ICP fitting and computed distances, not just viz.